### PR TITLE
[1/3] Support Open Telemetry (version bump, change interface)

### PR DESCRIPTION
**Pull request by chung-shopline**
This PR contains a lot of boilerplates:

- Changed most of the interfaces to take `context.Context`
- Bumped version to v2 since interface changed
  - may need to make a release for Go to recognize it
  - all `import`s are changed

This PR can be reviewed by quick glances as most changes are repetitive.

PR 2/3: https://github.com/chung-shopline/captin/pull/1
PR 3/3: https://github.com/chung-shopline/captin/pull/2

### CODE REVIEW

**Comment by chung-shopline** [comment]
- instead of keeping 2 versions of the interfaces, better abandon the old interface and bump major version
- trace circuit breaker also
- trace prioritiser also

File: internal/outgoing/custom.go [hunk 1]
```diff
@@ -22,12 +23,13 @@ func (c Custom) Sift(e *models.IncomingEvent, destinations []models.Destination,
 	for _, destination := range destinations {
 		eligible := true
 		for _, filter := range filters {
-			if eligible == false || filter.Applicable(*e, destination) == false {
+			if !filter.Applicable(ctx, *e, destination) {
```
Comment (chung-shopline) [inline internal/outgoing/custom.go:26]: split refactor into separate PR

File: internal/outgoing/dispatcher.go [hunk 2]
```diff
@@ -264,33 +267,37 @@ func (d *Dispatcher) processDelayedEvent(e models.IncomingEvent, timeRemain time
...
 		jsonString, jsonErr := json.Marshal(customizedPayload)
 		if jsonErr != nil {
 			panic(jsonErr)
 		}
+		ttl := dest.Config.GetThrottleValue() * 2
 		dLogger.WithFields(log.Fields{
 			"queueKey":       queueKey,
 			"event":          e,
 			"enqueuePayload": jsonString,
+			"ttl":            ttl,
```
Comment (chung-shopline) [inline internal/outgoing/dispatcher.go:281]: should split into another PR

File: senders/http_proxy_sender.go [hunk 3]
```diff
@@ -26,12 +27,7 @@ type HTTPProxyResponse struct {
 // in order to pass event meta data to destinations,
 // HTTPProxyEventSender only parses payload for general usage of 
 // third party API calls.
-type HTTPProxyEventSender struct {
```
Comment (chung-shopline) [inline senders/http_proxy_sender.go:29]: Should not delete this sender

File: test/internal/outgoing/dispatcher_test.go [hunk 4]
```diff
@@ -381,51 +383,52 @@ func TestDispatchEvents_Throttled_KeepThrottledDocuments_Multiple(t *testing.T)
...
+	sender.On("SendEvent", mock.Anything, mock.Anything, mock.Anything).Return(nil)
 
 	throttleID := "product.update.service_one.product_id-data"
 	throttleDocumentsID := "product.update.service_one.product_id-throttled_documents"
 
-	dispatcher.Dispatch(models.IncomingEvent{
+	// In Captain we are not reusing dispatcher for dispatcher.Dispatch
+	// (it should not be re-used anyway as there are states like Errors and targetDocument)
+	// If we reuse the same dispatcher in test case, it fails with data race, which won't happen in production
+	// So we copy the dispatcher (with the same stores)
```
Comment (chung-shopline) [inline test/internal/outgoing/dispatcher_test.go:398]: should comment on Dispatcher for no reuse

File: internal/outgoing/dispatcher.go [hunk 5]
```diff
@@ -86,14 +87,15 @@ func (d *Dispatcher) OnError(evt interfaces.IncomingEventInterface, err interfac
...
 		}).Error("Failed to dispatch event")
-		d.TriggerErrorHandler(dispatcherErr)
+		d.TriggerErrorHandler(ctx, dispatcherErr)
 	default:
 		dLogger.WithFields(log.Fields{"event": evt, "error": err}).Error("Unhandled error on dispatcher")
 	}
 }
 
 // Dispatch - Dispatch an event to outgoing webhook
 func (d *Dispatcher) Dispatch(
```
Comment (chung-shopline) [inline internal/outgoing/dispatcher.go:97]: Should add field and log for warning double dispatch

File: senders/http_proxy_sender.go (hunk 3 above)
Comment (chung-shopline) [reply senders/http_proxy_sender.go:29]: The sender is "revived" in another PR: `HTTPProxyEventSender`

**Review by neal-sl — APPROVED** [review]

**Review by thomaslishopline — APPROVED** [review]

**Review by neal-sl — APPROVED** [review]
